File: 백준/Gold/2473. 세 용액/세 용액.py
import sys
input = sys.stdin.readline

n = int(input())
num = list(map(int, input().split()))

def select_three():
    min_sum = sys.maxsize
    min_sum_list = []

    for i in range(n-2):
        left = i + 1
        right = len(num) - 1
        while left < right:
            num_sum = num[i] + num[left] + num[right]

            if abs(num_sum) < min_sum:
                min_sum = abs(num_sum)
                min_sum_list = [num[i], num[left], num[right]]

            if num_sum < 0:
                left += 1
            elif num_sum > 0:
                right -= 1
            else:
                return min_sum_list
            
    return min_sum_list

num.sort()
answer = select_three()
answer.sort()
print(*answer)

File: 백준/Gold/2473. 세 용액/세 용액.js
const input = require('fs').readFileSync(process.platform === 'linux' ? '/dev/stdin' : __dirname + '/input.txt').toString().trim().split('\n')

const N = Number(input[0])
const arr = input[1].split(' ').map(Number)
arr.sort((a, b) => a - b)

let best = Infinity
let answer = []
arr.forEach((fixed, idx) => {
  let left = idx + 1
  let right = N - 1

  while(left < right) {
    const sum = fixed + arr[left] + arr[right]
    
    if (Math.abs(sum) < best) {
      answer = [fixed, arr[left], arr[right]]
      best = Math.abs(sum)
    }

    if (sum < 0) left++ 
    else if (0 < sum) right--
    else break
  }
})

console.log(answer.join(' '))
